### PR TITLE
[HYD-789] Implement `pgxman auth login`

### PR DESCRIPTION
**Pull request by owenthereal**
Implement `pgxman auth login`. The command `pgxman auth` is hidden.

Demo:

```
$ pgxman auth login --registry http://localhost:8080/v1
Press Enter to log in to localhost:8080 in your browser...
Logged in as o@hydra.so
```

This requires https://github.com/pgxman/registry/pull/78

### CODE REVIEW

File: internal/auth/localserver.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+	w.Header().Add("content-type", "text/html")
+	if s.WriteSuccessHTML != nil {
+		s.WriteSuccessHTML(w)
+	} else {
+		defaultSuccessHTML(w)
+	}
+}
+
+func defaultSuccessHTML(w io.Writer) {
+	fmt.Fprintf(w, "<p>You may now close this page and return to the client app.</p>")
```
Comment (owenthereal) [inline internal/auth/localserver.go:89]: This could use some UI design later.

File: internal/auth/login.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+	if err != nil {
+		return err
+	}
+
+	user, err := client.GetUser(ctx)
+	if err != nil {
+		return err
+	}
+
+	if err := keyring.Set(keyringServiceName(opts.RegistryURL.Host), string(user.Email), token); err != nil {
```
Comment (owenthereal) [inline internal/auth/login.go:59]: Access token is stored in keyring using https://github.com/zalando/go-keyring

**Review by wuputah — APPROVED** [review]
I see this adds a token in USER_CONFIG_DIR via the change in `config.go`. Do we know how long the token lasts? Will the code currently re-authenticate automatically or is that something we still need to think about?

File: internal/auth/login.go (hunk 2 above)
Comment (wuputah) [reply internal/auth/login.go:59]: oh interesting. Zalando! ;)

**Comment by owenthereal** [comment]
> I see this adds a token in USER_CONFIG_DIR via the change in config.go. Do we know how long the token lasts? Will the code currently re-authenticate automatically or is that something we still need to think about.

The implementation currently reads client id and login endpoint from the `USER_CONFIG_DIR`: https://github.com/pgxman/pgxman/pull/164/files#diff-54c7c1af5fa8d5db4dc49f0e8e80e93ba2b1183ba4d5c9e2e5729e6deae6a3cdR17-R18. They are not secrets. I'm thinking. Eventually, we will hard-code them in code when the production environment is set up, while one could override in the config in `USER_CONFIG_DIR.` Happy to chat if there is a better option.

**Comment by owenthereal** [comment]
> Will the code currently re-authenticate automatically or is that something we still need to think about.

No, the client id and endpoint are "fix" info from the auth0 application. I guess you are thinking about whether access token will expire? Currently it doesn't, but we can configure in auth0 and the go oauth will deal with refresh automatically (pending to verify)

**Comment by wuputah** [comment]
OK. I'm also thinking ahead to solving the problem of how somehow might "login" in a non-interactive environment (Github Actions). I guess the user might need to get an API key or OAuth token for such a purpose.

**Comment by owenthereal** [comment]
>  I guess the user might need to get an API key or OAuth token for such a purpose.

Correct. We will need to introduce creating an access token on behalf of the user (similar to GitHub's PAT). This will require calling auth0's API in the registry to facilitate that.

**Pull request merged**